aws/resource_aws_cognito_user_pool.go

```diff
@@ -0,0 +1,902 @@
+		config, ok := configs[0].(map[string]interface{})
+
+		if !ok {
+			return errors.New("lambda_config is <nil>")
+		}
+
+		if config != nil {
+			params.LambdaConfig = expandCognitoUserPoolLambdaConfig(config)
+		}
+	}
+
+	if v, ok := d.GetOk("mfa_configuration"); ok {
+		params.MfaConfiguration = aws.String(v.(string))
+	}
+
+	policies := &cognitoidentityprovider.UserPoolPolicyType{}
+
+	if v, ok := d.GetOk("password_policy"); ok {
+		configs := v.([]interface{})
+		config, ok := configs[0].(map[string]interface{})
+
+		if !ok {
+			return errors.New("password_policy is <nil>")
+		}
+
+		if config != nil {
+			policies.PasswordPolicy = expandCognitoUserPoolPasswordPolicy(config)
+		}
+	}
+
+	params.Policies = policies
+
+	if v, ok := d.GetOk("schema"); ok {
+		configs := v.(*schema.Set).List()
+		params.Schema = expandCognitoUserPoolSchema(configs)
+	}
+
+	if v, ok := d.GetOk("sms_authentication_message"); ok {
+		params.SmsAuthenticationMessage = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("sms_configuration"); ok {
+		configs := v.([]interface{})
+		config, ok := configs[0].(map[string]interface{})
+
+		if !ok {
+			return errors.New("sms_configuration is <nil>")
+		}
+
+		if config != nil {
+			params.SmsConfiguration = expandCognitoUserPoolSmsConfiguration(config)
+		}
+	}
+
+	if v, ok := d.GetOk("username_attributes"); ok {
+		params.UsernameAttributes = expandStringList(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("verification_message_template"); ok {
+		configs := v.([]interface{})
+		config, ok := configs[0].(map[string]interface{})
+
+		if !ok {
+			return errors.New("verification_message_template is <nil>")
+		}
+
+		if config != nil {
+			params.VerificationMessageTemplate = expandCognitoUserPoolVerificationMessageTemplate(config)
+		}
+	}
+
+	if v, ok := d.GetOk("sms_verification_message"); ok {
+		params.SmsVerificationMessage = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("tags"); ok {
+		params.UserPoolTags = tagsFromMapGeneric(v.(map[string]interface{}))
+	}
+	log.Printf("[DEBUG] Creating Cognito User Pool: %s", params)
+
+	// IAM roles & policies can take some time to propagate and be attached
+	// to the User Pool
+	var resp *cognitoidentityprovider.CreateUserPoolOutput
+	err := resource.Retry(2*time.Minute, func() *resource.RetryError {
+		var err error
+		resp, err = conn.CreateUserPool(params)
+		if isAWSErr(err, "InvalidSmsRoleTrustRelationshipException", "Role does not have a trust relationship allowing Cognito to assume the role") {
+			log.Printf("[DEBUG] Received %s, retrying CreateUserPool", err)
+			return resource.RetryableError(err)
+		}
+		if isAWSErr(err, "InvalidSmsRoleAccessPolicyException", "Role does not have permission to publish with SNS") {
+			log.Printf("[DEBUG] Received %s, retrying CreateUserPool", err)
+			return resource.RetryableError(err)
+		}
+
+		return resource.NonRetryableError(err)
+	})
+	if err != nil {
+		return errwrap.Wrapf("Error creating Cognito User Pool: {{err}}", err)
```

Comment:
Really nitpick, but the errwrap seems unnecessary in this context since we don't really care about the original error type, it will be just printed to the user. Simple `fmt.Errorf` would do the job, IMO.

Reply:
Done!